ext/pgsql/pgsql.c

```diff
@@ -3211,18 +3289,20 @@ PHP_FUNCTION(pg_copy_from)
 					if (UNEXPECTED(!tmp)) {
 						return;
 					}
-					query = (char *)emalloc(ZSTR_LEN(tmp) + 2);
-					strlcpy(query, ZSTR_VAL(tmp), ZSTR_LEN(tmp) + 2);
-					if (ZSTR_LEN(tmp) > 0 && *(query + ZSTR_LEN(tmp) - 1) != '\n') {
-						strlcat(query, "\n", ZSTR_LEN(tmp) + 2);
+					zend_string *zquery = zend_string_alloc(ZSTR_LEN(tmp) + 1, false);
+					memcpy(ZSTR_VAL(zquery), ZSTR_VAL(tmp), ZSTR_LEN(tmp) + 1);
+					ZSTR_LEN(zquery) = ZSTR_LEN(tmp);
+					if (ZSTR_LEN(tmp) > 0 && ZSTR_VAL(zquery)[ZSTR_LEN(tmp)]  != '\n') {
+						ZSTR_VAL(zquery)[ZSTR_LEN(tmp)] = '\n';
+						ZSTR_LEN(zquery) ++;
 					}
-					if (PQputCopyData(pgsql, query, (int)strlen(query)) != 1) {
-						efree(query);
+					if (PQputCopyData(pgsql, ZSTR_VAL(zquery), ZSTR_LEN(tmp) > 0 ? ZSTR_LEN(tmp) + 1 : 0) != 1) {
```

Comment:
Isn't this what we want?

```suggestion
					if (PQputCopyData(pgsql, ZSTR_VAL(zquery), ZSTR_LEN(zquery)) != 1) {
```

Reply:
Yes that s the point of the length changes :) I forgot to update before going to sleep it seems.

Reply:
Relatable :P